sample/rules.go

```diff
@@ -127,7 +115,32 @@ func (s *RulesBasedSampler) GetSampleRate(trace *types.Trace) (rate uint, keep b
 			}
 		}
 
-		if matched == len(rule.Condition) {
+		if rule.Condition == nil || matched == len(rule.Condition) {
+			var rate uint
+			var keep bool
+
+			if rule.Downstream != nil {
+
+				var sampler Sampler
+				var found bool
+				if sampler, found = s.samplers[rule.String()]; !found {
```

Comment:
The mutex needs to be used to ensure thread safety when reading/writing with the map to prevent multiple threads reading an empty sampler and trying to create one then each trying to store it in the map. The creation of a sampler doesn't really need to be synchronised as it doesn't consume any external resources that need to be disposed of manually.

Instead, we could populate the samplers map during `Start`. This will separate the creation from usage and not needing to lazy load the samplers. We can do the error logs there too if we can't create a sampler.

Reply:
Moving all initialization to Start means we can just check if the sampler exists, and if not log an error and return `1, true`, negating the need to use a mutex (unless I'm missing something)

Reply:
Yeah, that's right. We wouldn't need the mutex at all. If the sampler doesn't exist, just return `1, true` 👍🏻 

If we're going to log for each failed attempt to get a sampler, it'll need to be debug.